lisp/x/profiler/callgrind.go

```diff
@@ -0,0 +1,248 @@
+	}
+	if p.writer == nil {
+		return errors.New("No output set in profiler")
+	}
+	fmt.Fprintf(p.writer, "version: 1\ncreator: elps %s (Go %s)\n", lisp.ElpsVersion, runtime.Version())
+	fmt.Fprintf(p.writer, "cmd: %s\npart: 1\npositions: line\n\n", "TODO")
+	fmt.Fprintf(p.writer, "events: Time_(ns) Memory_(bytes)\n\n")
+	p.callRefs = make(map[int32]*callRef)
+	p.startTime = time.Now()
+	p.refs = make(map[string]int)
+	p.refCounter = 0
+	p.enabled = true
+	p.Unlock()
+	p.incrementCallRef("ENTRYPOINT", &token.Location{
+		File: "-",
+		Path: "-",
+		Pos:  0,
+		Line: 0,
+		Col:  0,
+	})
+	return nil
+}
+
+func (p *callgrindProfiler) SetFile(filename string) error {
+	p.Lock()
+	defer p.Unlock()
+	if p.enabled {
+		return errors.New("Profiler already enabled")
+	}
+	pointer, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	p.writer = pointer
+	return nil
+}
+
+func (p *callgrindProfiler) Complete() error {
+	ref := p.getCallRefAndDecrement()
+	p.Lock()
+	defer p.Unlock()
+	// Generate entrypoint
+	ref.duration = time.Since(ref.start)
+	fmt.Fprintf(p.writer, "fl=%s\n", p.getRef(ref.file))
+	fmt.Fprintf(p.writer, "fn=%s\n", p.getRef(ref.name))
+	fmt.Fprintf(p.writer, "%d %d %d\n", 0, ref.duration, 0)
+	// Output the things we called
+	for _, entry := range ref.children {
+		fmt.Fprintf(p.writer, "cfl=%s\n", p.getRef(entry.file))
+		fmt.Fprintf(p.writer, "cfn=%s\n", p.getRef(entry.name))
+		fmt.Fprint(p.writer, "calls=1 0 0\n")
+		fmt.Fprintf(p.writer, "%d %d %d\n", entry.line, entry.duration, 0)
+	}
+	fmt.Fprint(p.writer, "\n")
+	duration := time.Now().Sub(p.startTime)
+	ms := &runtime.MemStats{}
+	runtime.ReadMemStats(ms)
+	fmt.Fprintf(p.writer, "summary %d %d\n\n", duration.Nanoseconds(), ms.TotalAlloc)
+	return p.writer.Close()
+}
+
+func (p *callgrindProfiler) getRef(name string) string {
+	if ref, ok := p.refs[name]; ok {
+		return fmt.Sprintf("(%d)", ref)
+	}
+	p.refCounter++
+	p.refs[name] = p.refCounter
+	return fmt.Sprintf("(%d) %s", p.refCounter, name)
+}
+
+func (p *callgrindProfiler) Start(function *lisp.LVal) {
+	if !p.enabled {
+		return
+	}
+	switch function.Type {
+	case lisp.LInt, lisp.LString, lisp.LFloat, lisp.LBytes, lisp.LError, lisp.LArray, lisp.LQuote, lisp.LNative, lisp.LQSymbol, lisp.LSortMap:
+		// We don't need to profile these types. We could, but we're not that LISP :D
+		return
+	case lisp.LFun, lisp.LSymbol, lisp.LSExpr:
+		// Mark the time and point of entry. It feels like we're building the call stack in Runtime
+		// again, but we're not - it's called, not callers.
+		_, _, name := p.getFunctionParameters(function)
+		p.incrementCallRef(name, function.Source)
+	default:
+		panic(fmt.Sprintf("Missing type %d", function.Type))
+	}
+}
+
+// Generates a call ref so the same item can be located again
+func (p *callgrindProfiler) incrementCallRef(name string, loc *token.Location) *callRef {
+	p.Lock()
+	defer p.Unlock()
+	var thread *int32
+	thread = &([]int32{1}[0])
```

Comment:
It's not that hard to get the thread ID with a little bit of cgo - I wrote the code for that and then discovered that I didn't need it as there's only one active thread in a Runtime at the moment. I left the container for the multiple threads in though as the overhead was minimal and it means it will be easy to put it back later.